code/modules/admin/battle_royale.dm

```diff
@@ -0,0 +1,363 @@
+	. = ..()
+	GLOB.enter_allowed = TRUE
+	world.update_status()
+
+//Trigger random events and shit, update the world border
+/datum/battle_royale_controller/process()
+	process_num ++
+	//Once every 50 seconds
+	if(prob(2))
+		generate_basic_loot(2)
+	//Once every 300 seconds
+	if(prob(0.333))
+		generate_good_drop()
+	var/living_victims = 0
+	var/winner
+	for(var/mob/living/M in players)
+		if(QDELETED(M))
+			players -= M
+			continue
+		if(M.x > 128 + field_wall.current_range + 1 && M.x < 128 - field_wall.current_range - 1 && M.y > 128 + field_wall.current_range + 1 && M.y < 128 - field_wall.current_range - 1)
+			M.gib()
+		if(!SSmapping.level_trait(M.z, ZTRAIT_STATION) && !SSmapping.level_trait(M.z, ZTRAIT_RESERVED))
+			to_chat(M, "<span class='warning'>You have left the z-level!</span>")
+			M.gib()
+		if(M.stat != DEAD)
+			living_victims ++
+			winner = M
+		CHECK_TICK
+	if(living_victims <= 1 && !debug_mode)
+		if(winner)
+			to_chat(world, "<span class='ratvar'><font size=18>[key_name(winner)] is the winner!</font></span>")
+			new /obj/item/melee/supermatter_sword(get_turf(winner))
+		qdel(src)
+		return
+	//Once every 15 seconsd
+	// 1,920 seconds (about 32 minutes per game)
+	if(process_num % (field_delay * field_jumps) == 0)
+		field_wall.current_range -= field_jumps
+		field_wall.recalculate_field(TRUE)
+	if(process_num > 1000 && prob(0.5))
+		generate_endgame_drop()
+
+//==================================
+// INITIALIZATION
+//==================================
+
+/datum/battle_royale_controller/proc/start()
+	toggle_ooc(FALSE)
+	to_chat(world, "<span class='ratvar'><font size=24>Battle Royale will begin soon...</span></span>")
+	//Stop new player joining
+	GLOB.enter_allowed = FALSE
+	world.update_status()
+	if(SSticker.current_state < GAME_STATE_PREGAME)
+		to_chat(world, "<span class=boldannounce>Battle Royale: Waiting for server to be ready...</span>")
+		SSticker.start_immediately = FALSE
+		UNTIL(SSticker.current_state >= GAME_STATE_PREGAME)
+		to_chat(world, "<span class=boldannounce>Battle Royale: Done!</span>")
+	//Delay pre-game if we are in it
+	if(SSticker.current_state == GAME_STATE_PREGAME)
+		//Force people to be not ready and start the game
+		for(var/mob/dead/new_player/player in GLOB.player_list)
+			to_chat(player, "<span class=greenannounce>You have been forced as an observer. When the prompt to join battle royale comes up, press yes. This is normal and you are still in queue to play.</span>")
+			player.ready = FALSE
+			player.make_me_an_observer(TRUE)
+		to_chat(world, "<span class='boldannounce'>Battle Royale: Force-starting game.</span>")
+		SSticker.start_immediately = TRUE
+	SEND_SOUND(world, sound('sound/misc/server-ready.ogg'))
+	sleep(50)
+	//Clear client mobs
+	to_chat(world, "<span class='boldannounce'>Battle Royale: Clearing world mobs.</span>")
+	for(var/mob/M as anything in GLOB.player_list)
+		if(isliving(M))
+			qdel(M)
+		CHECK_TICK
+	sleep(50)
```

Comment:
sleep is bad, why not use timer subsystem here?

Reply:
Will it feature the soon to be released rail gun as a drop?

Reply:
Sleep is better for linear animations, the timer subsystem can be unreliable at times when the server is under high load